src/System.Net.NameResolution/src/System/Net/NameResolutionPal.Unix.cs

```diff
@@ -110,134 +110,80 @@ private static unsafe IPHostEntry CreateHostEntry(Interop.libc.hostent* hostent)
 
         public static unsafe SocketError TryGetAddrInfo(string name, out IPHostEntry hostinfo, out int nativeErrorCode)
         {
-            var hints = new Interop.libc.addrinfo
+            Interop.Sys.HostEntry* entry = null;
+            int result = Interop.Sys.GetHostEntriesForName(name, &entry);
+            if (result != 0)
             {
-                ai_family = Interop.libc.AF_UNSPEC, // Get all address families
-                ai_flags = Interop.libc.AI_CANONNAME
+                hostinfo = NameResolutionUtilities.GetUnresolvedAnswer(name);
+                nativeErrorCode = result;
+                return GetSocketErrorForNativeError(result);
+            }
+
+            String canonicalName = Marshal.PtrToStringAnsi((IntPtr)entry->CanonicalName);
```

Comment:
All of this until the free'ing of entry should be in a try block, with the free'ing in a finally.